cdc/sink/codec/avro.go

```diff
@@ -193,41 +241,155 @@ func avroEncode(table *model.TableName, manager *AvroSchemaManager, tableVersion
+	mysql.TypeLong:       "INT",
+	mysql.TypeLonglong:   "BIGINT",
+	mysql.TypeFloat:      "FLOAT",
+	mysql.TypeDouble:     "DOUBLE",
+	mysql.TypeBit:        "BIT",
+	mysql.TypeNewDecimal: "DECIMAL",
+	mysql.TypeTinyBlob:   "TEXT",
+	mysql.TypeMediumBlob: "TEXT",
+	mysql.TypeBlob:       "TEXT",
+	mysql.TypeLongBlob:   "TEXT",
+	mysql.TypeVarchar:    "TEXT",
+	mysql.TypeVarString:  "TEXT",
+	mysql.TypeString:     "TEXT",
+	mysql.TypeEnum:       "ENUM",
+	mysql.TypeSet:        "SET",
+	mysql.TypeJSON:       "JSON",
+	mysql.TypeDate:       "DATE",
+	mysql.TypeDatetime:   "DATETIME",
+	mysql.TypeTimestamp:  "TIMESTAMP",
+	mysql.TypeDuration:   "TIME",
+	mysql.TypeYear:       "YEAR",
+}
+
+func getTiDBTypeFromColumn(col *model.Column) string {
+	tt := type2TiDBType[col.Type]
+	if col.Flag.IsUnsigned() && (tt == "INT" || tt == "BIGINT") {
+		return tt + " UNSIGNED"
+	}
+	if col.Flag.IsBinary() && tt == "TEXT" {
+		return "BLOB"
+	}
+	return tt
 }
 
 const (
-	timestampMillis logicalType = "timestamp-millis"
-	timeMillis      logicalType = "time-millis"
-	decimalType     logicalType = "decimal"
+	replacementChar = "_"
+	numberPrefix    = "_"
 )
 
-// ColumnInfoToAvroSchema generates the Avro schema JSON for the corresponding columns
-func ColumnInfoToAvroSchema(name string, columnInfo []*model.Column) (string, error) {
+// debezium-core/src/main/java/io/debezium/schema/FieldNameSelector.java
+// https://avro.apache.org/docs/current/spec.html#names
+func sanitizeName(name string) string {
+	changed := false
+	var sb strings.Builder
+	for i, c := range name {
+		if i == 0 && (c >= '0' && c <= '9') {
+			sb.WriteString(numberPrefix)
+			sb.WriteRune(c)
+			changed = true
+		} else if !(c == '_' ||
+			('a' <= c && c <= 'z') ||
+			('A' <= c && c <= 'Z') ||
+			('0' <= c && c <= '9')) {
+			sb.WriteString(replacementChar)
+			changed = true
+		} else {
+			sb.WriteRune(c)
+		}
+	}
+
+	sanitizedName := sb.String()
+	if changed {
+		log.Warn(
+			fmt.Sprintf(
+				"Name '%s' potentially not safe for serialization, replaced with '%s'",
+				name,
+				sanitizedName,
+			),
+		)
+	}
+	return sanitizedName
+}
+
+// escape ".", it may has special meanings for sink connectors
+func sanitizeTopic(name string) string {
+	return strings.ReplaceAll(name, ".", "_")
+}
+
+// https://github.com/debezium/debezium/blob/9f7ede0e0695f012c6c4e715e96aed85eecf6b5f \
+// /debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/antlr/ \
+// MySqlAntlrDdlParser.java#L374
+func escapeEnumAndSetOptions(option string) string {
+	option = strings.ReplaceAll(option, ",", "\\,")
+	option = strings.ReplaceAll(option, "\\'", "'")
+	option = strings.ReplaceAll(option, "''", "'")
+	return option
+}
+
+func getAvroNamespace(namespace string, tableName *model.TableName) string {
+	return sanitizeName(namespace) + "." + sanitizeName(tableName.Schema)
+}
+
+type avroSchema struct {
+	Type       string            `json:"type"`
+	Parameters map[string]string `json:"connect.parameters"`
```

Comment:
``json:"connect.parameters"`` Is this a typo? Or do we have to write it that way?